java/arcs/core/storage/util/SimpleQueue.kt

```diff
@@ -0,0 +1,69 @@
+package arcs.core.storage.util
+
+import kotlin.coroutines.coroutineContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * A very basic implementation of [OperationQueue]. It dispatches jobs on coroutines in the
+ * coroutine context of callers of the [enqueue] method.
+ *
+ * If an [onEmpty] method is provided at construction, that method will run at the end of
+ * any drain cycle.
+ */
+class SimpleQueue(
+    private val onEmpty: (suspend () -> Unit)? = null
+) : OperationQueue {
+
+    private val mutex = Mutex()
+    @OptIn(ExperimentalStdlibApi::class)
+    private val queue = ArrayDeque<suspend () -> Unit>()
+    private var drainJob: Job? = null
+
+    /**
+     * Places the provided block on the internal operation queue.
+     *
+     * If there is not a coroutine currently draining the queue, one will be started, using
+     * the coroutineContext of the caller.
+     *
+     * It's safe to call this method from any thread/coroutine.
+     */
+    override suspend fun enqueue(op: Op) = mutex.withLock {
+        // Update the queue, and at the same time, check if any drain coroutines are running.
+        queue += op
+
+        if (drainJob == null) {
+            drainJob = CoroutineScope(coroutineContext).launch {
```

Comment:
@jasonwyatt I'm stumped about what to do here.

We need the function to be `suspend`, because it performs operations within a `mutex.withLock`. That means we can't make it `Coroutine.enqueue`, since then there's ambiguity about what `coroutineContext` is. 

The article Yuan referenced in your docs also claims that a suspend method should never launch a job. 

What do you think we should be doing here?